### PR TITLE
Validate urls in parallel

### PR DESCRIPTION
**Pull request by AndrewVos**
We are getting timeouts in integration if all the urls are requested
synchronously.

### CODE REVIEW

File: app/models/govspeak_url_checker.rb [hunk 1]
```diff
@@ -4,15 +4,22 @@ def initialize govspeak
...
+    end
 
-      !checked_url.ok?
+    broken_urls = Parallel.map(urls_to_check, in_threads: 10) do |u|
+      check_url(u) ? nil : u
+    end.compact
+
+    urls_to_check.each do |url|
+      CheckedUrl.find_or_initialize_by(url: url) do |checked_url|
+        checked_url.ok = !broken_urls.include?(checked_url.url)
```
Comment (benlovell) [inline app/models/govspeak_url_checker.rb]: You could use `exclude?` to avoid the negation.

File: app/models/govspeak_url_checker.rb [hunk 2]
```diff
@@ -4,15 +4,22 @@ def initialize govspeak
...
-        checked_url.save!
-      end
+    urls_to_check = extract_all_urls_from_govspeak.select do |url|
+      url = CheckedUrl.find_or_initialize_by(url: url)
+      url.new_record? || url.expired?
+    end
 
-      !checked_url.ok?
+    broken_urls = Parallel.map(urls_to_check, in_threads: 10) do |u|
+      check_url(u) ? nil : u
```
Comment (benlovell) [inline app/models/govspeak_url_checker.rb]: Did you first try changing `#check_url` to perform a `HEAD` request? This is probably piles more efficient.

File: app/models/govspeak_url_checker.rb [hunk 3]
```diff
@@ -4,44 +4,63 @@ def initialize govspeak
...
+  def check_urls urls
+    broken_urls = []
+    hydra = Typhoeus::Hydra.new(max_concurrency: 5)
+
+    urls.each do |url|
       options = {
-        follow_redirects: true,
-        timeout: 5,
+        followlocation: true,
+        forbid_reuse: true,
```
Comment (thehenster) [inline app/models/govspeak_url_checker.rb]: Does this do anything when each request is in a separate thread? I couldn't imagine it managing to keep persistent connections across threads for some reason..

File: app/models/govspeak_url_checker.rb [hunk 4]
```diff
@@ -4,44 +4,63 @@ def initialize govspeak
...
       end
 
       begin
-        HTTParty.get(
-          url,
-          options,
-        ).ok?
-      rescue
-        false
+        Addressable::URI.parse(url)
```
Comment (thehenster) [inline app/models/govspeak_url_checker.rb]: Could we add addressable to the gemfile if we're gonna use it over normal URI? It looks like it's just a fortunate dependency of other gems atm.

File: app/models/govspeak_url_checker.rb [hunk 5]
```diff
@@ -4,44 +4,63 @@ def initialize govspeak
...
+      url = CheckedUrl.find_or_initialize_by(url: url)
+      url.new_record? || url.expired?
+    end
 
-      !checked_url.ok?
+    if urls_to_check.any?
+      broken_urls = check_urls(urls_to_check)
+      urls_to_check.each do |url|
+        CheckedUrl.find_or_initialize_by(url: url) do |checked_url|
+          checked_url.ok = !broken_urls.include?(checked_url.url)
```
Comment (thehenster) [inline app/models/govspeak_url_checker.rb]: Ben's `exclude?` suggestion would be nice.

File: app/models/govspeak_url_checker.rb [hunk 6]
```diff
@@ -4,44 +4,63 @@ def initialize govspeak
...
+        end.save!
```
Comment (thehenster) [inline app/models/govspeak_url_checker.rb]: If saving immediately we could use http://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-find_or_create_by-21 ?

File: app/models/govspeak_url_checker.rb [hunk 7]
```diff
@@ -4,44 +4,63 @@ def initialize govspeak
...
+      []
     end
   end
 
   private
 
-    def check_url url
+  def check_urls urls
+    broken_urls = []
+    hydra = Typhoeus::Hydra.new(max_concurrency: 5)
```
Comment (thehenster) [inline app/models/govspeak_url_checker.rb]: Does it spew the same error regarding AR connections if we go above 5?

File: app/models/govspeak_url_checker.rb (hunk 3 above)
Comment (AndrewVos) [inline app/models/govspeak_url_checker.rb]: Yes it's curl that handles the parallel processing afaik

File: app/models/govspeak_url_checker.rb (hunk 4 above)
Comment (AndrewVos) [inline app/models/govspeak_url_checker.rb]: Done

**Pull request merged**